wpilibc/src/main/native/include/frc/controller/NonlinearPlantInversionFeedforward.h

```diff
@@ -0,0 +1,158 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2020 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <array>
+#include <functional>
+
+#include <Eigen/Core>
+#include <units/units.h>
+
+#include "frc/system/NumericalJacobian.h"
+
+namespace frc {
+
+template <int N>
+using Vector = Eigen::Matrix<double, N, 1>;
+
+/**
+ * Constructs a plant inversion model-based feedforward from given model
+ * dynamics.
+ *
+ * <p>B is calculated through a {@link
+ * edu.wpi.first.wpilibj.system.NumericalJacobian}.
+ *
+ * <p>The feedforward is calculated as
+ * u_ff = B<sup>+</sup> (rDot - f(x)), were B<sup>+</sup> is the pseudoinverse
+ * of B.
+ *
+ * <p>For more on the underlying math, read
+ * https://file.tavsys.net/control/controls-engineering-in-frc.pdf.
+ */
+template <int States, int Inputs>
+class NonlinearPlantInversionFeedforward {
+ public:
+  /**
+   * Constructs a feedforward with given model dynamics.
+   *
+   * @param f         A vector-valued function of x, the state, and
+   *                  u, the input, that returns the derivative of
+   *                  the state vector.
+   * @param dtSeconds The timestep between calls of calculate().
+   */
+  NonlinearPlantInversionFeedforward(
+      std::function<Vector<States>(const Vector<States>&,
+                                   const Vector<Inputs>&)>
+          f,
+      units::second_t dt)
+      : m_dt(dt), m_f(f) {
+    m_B = NumericalJacobianU<States, States, Inputs>(f, Vector<States>::Zero(),
```

Comment:
I guess this isn't technically correct. One could provide a model whose input matrix varies with the state. To fix this, we'd have to resample B every timestep based on r, but that's paying for something we don't use much, if at all, in FRC.

There's really three classes of models we care about: linear (`dx/dt = Ax + Bu`), control-affine (`dx/dt = f(x) + Bu`), and nonlinear (`dx/dt = f(x, u)`). For FRC, we arguably only need the first two. To support control-affine, we can either do what we're doing now with taking f(x, u) and determining B implicitly, or take a function of just the state and a separate B. I think what we're doing now is most convenient, and the extra Jacobian in the ctor isn't a big deal.

Therefore, I suggest we rename this class to `ControlAffinePlantInversionFeedforward` and keep the implementation as is.